src/map.h

```diff
@@ -4,69 +4,82 @@
  */
 
 /*
- * C Implementation for C++ std::map using red-black tree.
- *
- * Any data type can be stored in a map, just like std::map.
- * A map instance requires the specification of two file types:
- *   1. the key;
- *   2. what data type the tree node will store;
- *
- * It will also require a comparison function to sort the tree.
+ * This map implementation is heavily modified based on the rb.h header file of
+ * jemalloc. The original rb.h file from jemalloc has been extensively adapted
+ * and tailored for this map implementation. Credit and thanks to jemalloc for
+ * providing the initial implementation and inspiration.
  */
 
 #pragma once
 
 #include <stdbool.h>
 #include <stddef.h>
+#include <stdint.h>
+#include <stdlib.h>
+
+#define RB_MAX_DEPTH (sizeof(void *) << 4)
+
+/* Node structure */
+#define rb_node(x_type)           \
+    struct {                      \
+        x_type *left, *right_red; \
+    }
+
+typedef struct map_node map_node_t;
+struct map_node {
+    void *key;
+    void *val;
+    rb_node(map_node_t) link;
+};
+
+typedef struct rb_tree {
+    map_node_t *root;
+    int (*cmp)(const map_node_t *, const map_node_t *);
+    size_t key_size;
+    size_t val_size;
+} map_internal_t;
+
+typedef map_internal_t *map_t;
+
+typedef struct {
+    map_node_t *prev, *node;
+    size_t count;
+} map_iter_t;
+
+#define map_iter_value(it, type) (*(type *) (it)->node->val)
 
 enum { _CMP_LESS = -1, _CMP_EQUAL = 0, _CMP_GREATER = 1 };
 
 /* Integer comparison */
-static inline int map_cmp_int(const void *arg0, const void *arg1)
+static inline int map_cmp_int(const map_node_t *arg0, const map_node_t *arg1)
```

Comment:
Can you check if the use of `map_node_t` violates the casting rules? That is, we shall rely on the casting from `void *`.